WalletWasabi/Stores/IndexStore.cs

```diff
@@ -192,204 +121,107 @@ private bool TryProcessFilterNoLock(FilterModel filter, bool enqueue)
 
 	public async Task AddNewFiltersAsync(IEnumerable<FilterModel> filters)
 	{
-		var successAny = false;
-
-		foreach (var filter in filters)
+		if (NewFilter is null)
 		{
-			var success = false;
-
-			using (await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
-			{
-				success = TryProcessFilterNoLock(filter, enqueue: true);
-			}
+			// Lock once.
+			using IDisposable lockDisposable = await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false);
 
-			successAny = successAny || success;
+			using SqliteTransaction sqliteTransaction = IndexStorage.BeginTransaction();
 
-			if (success)
+			foreach (FilterModel filter in filters)
 			{
-				NewFilter?.Invoke(this, filter); // Event always outside the lock.
+				// TODO: Should throw?
+				_ = TryProcessFilterNoLock(filter, enqueue: true);
 			}
-		}
 
-		if (successAny)
-		{
-			AbandonedTasks.AddAndClearCompleted(TryCommitToFileAsync(TimeSpan.FromSeconds(3)));
-		}
-	}
-
-	public async Task<FilterModel> RemoveLastFilterAsync()
-	{
-		FilterModel? filter = null;
-
-		using (await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
-		{
-			filter = ImmatureFilters.Last();
-			ImmatureFilters.RemoveLast();
-			if (SmartHeaderChain.TipHeight != filter.Header.Height)
-			{
-				throw new InvalidOperationException($"{nameof(SmartHeaderChain)} and {nameof(ImmatureFilters)} are not in sync.");
-			}
-			SmartHeaderChain.RemoveTip();
+			sqliteTransaction.Commit();
```

Comment:
What happen in case of  `TryProcessFilterNoLock` fails with exception? Should we rollback?

Reply:
Disposal of an incomplete transaction should automatically roll back it in ADO.NET, but that part is implemented in the driver's code.